# Overview
Data transformation is a process of modifying and converting raw data into a clean, structured, and organized format that is ready for further analysis and modeling. This process can involve several different steps such as merging data from multiple sources, reshaping the data to make it consistent, and grouping data to help reveal patterns and insights. The end result is a dataset that is standardized, optimized, and well-suited for statistical modeling and other types of data analysis.

The importance of data transformation in data wrangling lies in the fact that it ensures that the data is prepared in a format that can be easily consumed and analyzed. Raw data is often in a format that is not suitable for analysis and modeling. Data transformation helps to reshape the data into a format that is easier to work with and better suited for further analysis.

Data transformation helps to merge and integrate data from different sources into a single, unified data set. This helps to overcome the limitations of working with isolated data sources and enables more comprehensive and accurate analysis. Additionally, data transformation enables aggregation of the data, which involves summarizing and reducing the data to a more manageable and interpretable size. This is particularly useful when working with large datasets.

In this module, we will cover the following topics:

I. Merging data: Processes for combining data from multiple sources into a single, unified dataset.
II. Aggregating data: Methods for summarizing and grouping data based on specified criteria.
III. Reshaping data: Techniques for restructuring and rearranging data into desired forms.

# Learning Objectives
In this module, the learners will:

* Understand how to convert raw data into a clean and structured format
* Synthesize information from multiple datasets to create a unified dataset
* Summarize and interpret data to produce valuable insights
* Evaluate the quality of data to facilitate effective decision-making
Let's get started!

# Dataset
## Titanic dataset
This is a well-known and widely used dataset in the field of data analysis and machine learning. This dataset contains information about the passengers on the Titanic ship, including their demographic information, ticket information, and survival status. In this exercise, we're using the Titanic dataset and its two sub-datasets (passengers_personal_info and travel_info) to demonstrate the merging, aggregating, and reshaping of data through Pandas functions.

Here's a description of the columns in the dataset:

* PassengerId: This column is a unique identifier assigned to each passenger.
* Age: This column specifies the age of the passenger.
* Name: This column specifies the name of the passenger.
* Sex: This column specifies the gender of the passenger (Male or Female).
* Survived: This column specifies whether the passenger survived the Titanic disaster or not. The values in this column can either be 0 (did not survive) or 1 (survived).
* Pclass (Passenger Class): This column specifies the class of the passenger (1st, 2nd, or 3rd class).
* SibSp (Siblings/Spouses Aboard): This column specifies the number of siblings or spouses the passenger was traveling with.
* Parch (Parents/Children Aboard): This column specifies the number of parents or children the passenger was traveling with.
* Ticket: This column specifies the ticket number assigned to the passenger.
* Fare: This column specifies the fare paid by the passenger for their ticket.
* Cabin: This column specifies the cabin number assigned to the passenger.
* Embarked: This column specifies the port where the passenger boarded the Titanic (C = Cherbourg; Q = Queenstown; S = Southampton).

## Loading datasets
This code imports the Pandas library and reads a CSV file that contains the personal information of Titanic passengers. The dataset includes information about the passenger id, name, age, and sex.

In [25]:
# Import the Pandas library
import pandas as pd

# Use the 'pd.read_csv' method to read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")
print(passengers_personal_info)

   PassengerId   Age                                               Name  \
0            1  22.0                            Braund, Mr. Owen Harris   
1            2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3  26.0                             Heikkinen, Miss. Laina   
3            4  35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            6   NaN                                   Moran, Mr. James   
5            7  54.0                            McCarthy, Mr. Timothy J   
6            8   2.0                     Palsson, Master. Gosta Leonard   
7           10  14.0                Nasser, Mrs. Nicholas (Adele Achem)   
8           11   4.0                    Sandstrom, Miss. Marguerite Rut   

      Sex  
0    male  
1  female  
2  female  
3  female  
4    male  
5    male  
6    male  
7  female  
8  female  


This code reads a CSV file named "travel_info.csv", which contains the travel information of Titanic passengers. This dataset includes Survived, Pclass, SibSp, Parch, Ticket, Fare, Cabin, and Embarked columns.

In [26]:
# Import the Pandas library
import pandas as pd

# Use the 'pd.read_csv' method to read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")
print(travel_info)

   PassengerId  Survived  Pclass  SibSp  Parch            Ticket     Fare  \
0            1         0       3      1      0         A/5 21171   7.2500   
1            2         1       1      1      0          PC 17599  71.2833   
2            3         1       3      0      0  STON/O2. 3101282   7.9250   
3            5         0       3      0      0            373450   8.0500   
4            6         0       3      0      0            330877   8.4583   
5            7         0       1      0      0             17463  51.8625   
6            8         0       3      3      1            349909  21.0750   
7            9         1       3      0      2            347742  11.1333   
8           11         1       3      1      1           PP 9549  16.7000   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3   NaN        S  
4   NaN        Q  
5   E46        S  
6   NaN        S  
7   NaN        S  
8    G6        S  


## What is merging data?
The process of merging data is a common task in data analysis and data science, as it allows you to combine data from multiple sources into a single data structure for analysis and modeling. This process can be performed in different ways, depending on the type of join operation to be performed.

## Why is it important?
One common use case is analyzing customer behavior for businesses. By merging customer transaction and demographic data, businesses can gain insights into their customers' behavior, preferences, and needs to inform marketing strategies, product development, and customer service initiatives.

In summary, merging data is a crucial step in many data analysis tasks. By combining data from multiple sources, businesses can gain deeper insights into their customers and make data-driven decisions to improve their overall performance.

## Performing inner join
An inner join only includes rows that have matching values in both datasets. To perform an inner join, we use the 'merge()' function which combines two dataframes into a single one, based on a common column like 'PassengerId'.

The resulting merged dataset contains only the rows where there is a match, effectively creating a subset of the original datasets.

In [27]:
# Import the Pandas library
import pandas as pd 

# # Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Perform an inner join on the 'PassengerId' column
# Only the rows that have matching values in both datasets will be included in the result
inner_join = pd.merge(passengers_personal_info, travel_info, on='PassengerId', how='inner')
print(inner_join)

   PassengerId   Age                                               Name  \
0            1  22.0                            Braund, Mr. Owen Harris   
1            2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3  26.0                             Heikkinen, Miss. Laina   
3            6   NaN                                   Moran, Mr. James   
4            7  54.0                            McCarthy, Mr. Timothy J   
5            8   2.0                     Palsson, Master. Gosta Leonard   
6           11   4.0                    Sandstrom, Miss. Marguerite Rut   

      Sex  Survived  Pclass  SibSp  Parch            Ticket     Fare Cabin  \
0    male         0       3      1      0         A/5 21171   7.2500   NaN   
1  female         1       1      1      0          PC 17599  71.2833   C85   
2  female         1       3      0      0  STON/O2. 3101282   7.9250   NaN   
3    male         0       3      0      0            330877   8.4583   NaN   
4    male

In the above result of the inner join, certain cells have been skipped. This is because some of the values for the 'PassengerId' column are present in the left dataset ('passengers_personal_info') but not in the right dataset ('travel_info'), and vice versa.

## Performing left join
A left join returns all the rows from the left dataset (in this case, the  'passengers_personal_info' dataset) and only the matching rows from the right dataset (in this case, the 'ticket_info' dataset). If there are no matching rows in the right dataset, the result will have NaN values for the columns from the right dataset.

In [28]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Perform a left join on the 'PassengerId' column
left_join = pd.merge(passengers_personal_info, travel_info, on='PassengerId', how='left')
print(left_join)

   PassengerId   Age                                               Name  \
0            1  22.0                            Braund, Mr. Owen Harris   
1            2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3  26.0                             Heikkinen, Miss. Laina   
3            4  35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            6   NaN                                   Moran, Mr. James   
5            7  54.0                            McCarthy, Mr. Timothy J   
6            8   2.0                     Palsson, Master. Gosta Leonard   
7           10  14.0                Nasser, Mrs. Nicholas (Adele Achem)   
8           11   4.0                    Sandstrom, Miss. Marguerite Rut   

      Sex  Survived  Pclass  SibSp  Parch            Ticket     Fare Cabin  \
0    male       0.0     3.0    1.0    0.0         A/5 21171   7.2500   NaN   
1  female       1.0     1.0    1.0    0.0          PC 17599  71.2833   C85   
2  female      

The code above performs a left join using the 'PassengerId' column as the join key. All rows from the 'passengers_personal_info' dataset will be included in the result, and any missing values from the 'travel_info' dataset will be filled with NaN.

## Performing right join
Similarly, a right join returns all the rows from the right dataset (in this case, the 'ticket_info' dataset) and the matching rows from the left dataset (in this case, the 'passengers_personal_info'). If there are no matching rows in the left dataset, the result will have NaN values for the columns from the left dataset.

In [29]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Perform a right join on the 'PassengerId' column
right_join = pd.merge(passengers_personal_info, travel_info, on='PassengerId', how='right')
print(right_join)

   PassengerId   Age                                               Name  \
0            1  22.0                            Braund, Mr. Owen Harris   
1            2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3  26.0                             Heikkinen, Miss. Laina   
3            5   NaN                                                NaN   
4            6   NaN                                   Moran, Mr. James   
5            7  54.0                            McCarthy, Mr. Timothy J   
6            8   2.0                     Palsson, Master. Gosta Leonard   
7            9   NaN                                                NaN   
8           11   4.0                    Sandstrom, Miss. Marguerite Rut   

      Sex  Survived  Pclass  SibSp  Parch            Ticket     Fare Cabin  \
0    male         0       3      1      0         A/5 21171   7.2500   NaN   
1  female         1       1      1      0          PC 17599  71.2833   C85   
2  female      

The code above performs a right join using the 'PassengerId' column as the join key. All rows from the 'travel_info' dataset will be included in the result, and any missing values from the 'passengers_personal_info' dataset will be filled with NaN.

## Performing outer join
An outer join returns all the rows from both datasets, with matching rows combined and non-matching rows filled with NaN values for the columns that do not have matching data.


In [30]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv") 

# Perform an outer join on the 'PassengerId' column

# All rows from both datasets will be included in the result, and any 

# missing values will be filled with NaN
outer_join = pd.merge(passengers_personal_info, travel_info, on='PassengerId', how='outer')
print(outer_join)

    PassengerId   Age                                               Name  \
0             1  22.0                            Braund, Mr. Owen Harris   
1             2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2             3  26.0                             Heikkinen, Miss. Laina   
3             4  35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4             6   NaN                                   Moran, Mr. James   
5             7  54.0                            McCarthy, Mr. Timothy J   
6             8   2.0                     Palsson, Master. Gosta Leonard   
7            10  14.0                Nasser, Mrs. Nicholas (Adele Achem)   
8            11   4.0                    Sandstrom, Miss. Marguerite Rut   
9             5   NaN                                                NaN   
10            9   NaN                                                NaN   

       Sex  Survived  Pclass  SibSp  Parch            Ticket     Fare Cabin  \
0     ma

In the result above, the order of the rows with null values can change depending on the order of the datasets in the join. You can use the 'sort' parameter in the Pandas 'merge()' function to specify the sorting behavior of the merged result based on the join key.

In [31]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv") 

# Perform an outer join on the 'PassengerId' column

# All rows from both datasets will be included in the result, and any 

# Perform an outer join on the 'PassengerId' column with sort parameter
outer_join = pd.merge(passengers_personal_info, travel_info, on='PassengerId', how='outer', sort=True)
print(outer_join)

    PassengerId   Age                                               Name  \
0             1  22.0                            Braund, Mr. Owen Harris   
1             2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2             3  26.0                             Heikkinen, Miss. Laina   
3             4  35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4             5   NaN                                                NaN   
5             6   NaN                                   Moran, Mr. James   
6             7  54.0                            McCarthy, Mr. Timothy J   
7             8   2.0                     Palsson, Master. Gosta Leonard   
8             9   NaN                                                NaN   
9            10  14.0                Nasser, Mrs. Nicholas (Adele Achem)   
10           11   4.0                    Sandstrom, Miss. Marguerite Rut   

       Sex  Survived  Pclass  SibSp  Parch            Ticket     Fare Cabin  \
0     ma

## Note

In Pandas, you can perform join operations using either the 'join()' function or the 'merge()' function. While 'join()' is a convenient wrapper for common join operations, 'merge()' offers more flexibility and control. It's recommended to use 'merge()' for most join operations, and 'join()' only when joining on indices instead of columns.


## Performing cross join
A cross-join, also known as a cartesian product, is a join operation. It returns a result set that includes every possible combination of rows from both tables, where each row from the first table is paired with each row from the second table.

Here's an example of performing a cross-join between 'passengers_personal_info' and 'travel_info' datasets:

In [32]:
# Import the Pandas library
import pandas as pd 

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# The 'assign' method is used to add a new column "key" to both DataFrames.
# This is necessary because the merge method requires a common key to perform 
# the join operation. Although both DataFrames have a column called 
# PassengerId, this column cannot be used as the common key because the 
# values are not unique across the two DataFrames. 

# In this case, the column is assigned a constant value of 1 for all rows.

# Finally, the "key" column is dropped from the resulting DataFrame using the drop method. 
# This is done because the column was only added as a means to perform the join, and 
# it is no longer needed in the final DataFrame.
cross_join = passengers_personal_info.assign(key=1).merge(travel_info.assign(key=1), on='key', how='outer').drop('key', axis=1)
print(cross_join)

    PassengerId_x   Age                             Name     Sex  \
0               1  22.0          Braund, Mr. Owen Harris    male   
1               1  22.0          Braund, Mr. Owen Harris    male   
2               1  22.0          Braund, Mr. Owen Harris    male   
3               1  22.0          Braund, Mr. Owen Harris    male   
4               1  22.0          Braund, Mr. Owen Harris    male   
..            ...   ...                              ...     ...   
76             11   4.0  Sandstrom, Miss. Marguerite Rut  female   
77             11   4.0  Sandstrom, Miss. Marguerite Rut  female   
78             11   4.0  Sandstrom, Miss. Marguerite Rut  female   
79             11   4.0  Sandstrom, Miss. Marguerite Rut  female   
80             11   4.0  Sandstrom, Miss. Marguerite Rut  female   

    PassengerId_y  Survived  Pclass  SibSp  Parch            Ticket     Fare  \
0               1         0       3      1      0         A/5 21171   7.2500   
1               2      

The resulting DataFrame contains all possible combinations of rows from 'passengers_personal_info' and 'travel_info'. Note that if the two input DataFrames have 'n' rows each, the resulting DataFrame will have 'n' rows x 'n' rows = 'n^2' rows.


## Performing concatenation
Concatenation is the process of combining multiple datasets into a single dataset. In the case of the Titanic dataset, concatenation can be useful when you have separate datasets for each class or each embarkation port and you want to combine them into a single dataset for analysis.

Let's see the result when we combine the 'passengers_personal_info' and 'travel_info' datasets.

In [33]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Concatenate the two datasets
# The 'axis' parameter determines the axis along which to concatenate the dataframes
titanic_concat = pd.concat([passengers_personal_info, travel_info], axis=1)

# Preview the resulting dataset
print(titanic_concat)

   PassengerId   Age                                               Name  \
0            1  22.0                            Braund, Mr. Owen Harris   
1            2  38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2            3  26.0                             Heikkinen, Miss. Laina   
3            4  35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4            6   NaN                                   Moran, Mr. James   
5            7  54.0                            McCarthy, Mr. Timothy J   
6            8   2.0                     Palsson, Master. Gosta Leonard   
7           10  14.0                Nasser, Mrs. Nicholas (Adele Achem)   
8           11   4.0                    Sandstrom, Miss. Marguerite Rut   

      Sex  PassengerId  Survived  Pclass  SibSp  Parch            Ticket  \
0    male            1         0       3      1      0         A/5 21171   
1  female            2         1       1      1      0          PC 17599   
2  female            

In this example, 'axis = 1' refers to concatenating dataframes horizontally, along the columns. This means the columns of one dataframe are added to the columns of the other dataframe.

When 'axis = 0' is used in the 'pd.concat()' function, the dataframes are concatenated vertically. The result is the rows of one dataframe followed by the rows of the other dataframe.

### Note 
It is worth noting that concatenation can be performed on any set of columns in the Titanic dataset, not just the embarkation port and class columns. The choice of which columns to concatenate depends on the hypothesis being explored and the availability of data.

# Aggregating Data

## What is aggregating data?
Aggregation is an essential part of data transformation, as it helps simplify and summarize complex data, making it easier to understand and analyze. This can be done through techniques such as grouping, summarizing, and calculating statistical measures, which provide a clearer picture of the data being analyzed.

## Why is it important?
Without aggregation, analyzing large and complex data sets would be a challenging and time-consuming task. By simplifying and summarizing data, aggregation enables analysts and data scientists to identify patterns, trends, and insights that may not be apparent from examining individual data points.

One use case for aggregation is in analyzing website traffic data. Websites typically generate a large amount of data, including user behavior, page views, and referral sources. By aggregating website traffic data, analysts can identify peak traffic times, popular pages, and referral sources that generate the most traffic. This information can be used to optimize website design and content, improve user experience, and increase engagement and conversions.

## Grouping data
Grouping is a technique in data transformation used to aggregate and summarize data based on a specific attribute or feature. We will apply grouping in the 'travel_info' dataset to explore relationships within the data and make it easier to analyze and understand.

Here is an example of how you can use grouping on the 'travel_info' dataset to count the number of passengers who survived:

In [34]:
# Import the Pandas library
import pandas as pd

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# The groupby function is used to group rows of a DataFrame based on the values in 'Survived' column
grouped = travel_info.groupby('Survived')

# Count the number of passengers who survived
survived = grouped['PassengerId'].count()
print(survived)

Survived
0    5
1    4
Name: PassengerId, dtype: int64


The resulting survived dataframe will contain the number of passengers who survived (value of 1) and those who did not survive (value of 0).

### NOTE
In the provided code, the 'count()' function is applied to the 'PassengerId' column to count the number of passengers who survived in each group. However, other aggregating functions such as mean, sum, median, min, or max can also be used based on the hypothesis being investigated. It is essential to choose the appropriate aggregating function and ensure that the data being analyzed is appropriate for the chosen function.


## Binning data
Binning is a technique to group continuous or large-scale data into smaller, more manageable, meaningful intervals or "bins". This process helps reduce the noise and complexity of the data, making it easier to visualize and understand.

The following code creates a new column 'Fare_Range' in the 'travel_info' dataframe by dividing the 'Fare' column into ranges of 10:

In [35]:
# Import the Pandas library
import pandas as pd 

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Create a new column 'Fare_Range' by applying binning on the 'Fare' column
travel_info['Fare_Range'] = pd.cut(travel_info['Fare'], bins=[0, 10, 20, 30, 40, 50, 60, 70, travel_info['Fare'].max()], labels=['0-10', '10-20', '20-30', '30-40', '40-50', '50-60', '60-70', '70-80'])

# Verify the result
print(travel_info)

   PassengerId  Survived  Pclass  SibSp  Parch            Ticket     Fare  \
0            1         0       3      1      0         A/5 21171   7.2500   
1            2         1       1      1      0          PC 17599  71.2833   
2            3         1       3      0      0  STON/O2. 3101282   7.9250   
3            5         0       3      0      0            373450   8.0500   
4            6         0       3      0      0            330877   8.4583   
5            7         0       1      0      0             17463  51.8625   
6            8         0       3      3      1            349909  21.0750   
7            9         1       3      0      2            347742  11.1333   
8           11         1       3      1      1           PP 9549  16.7000   

  Cabin Embarked Fare_Range  
0   NaN        S       0-10  
1   C85        C      70-80  
2   NaN        S       0-10  
3   NaN        S       0-10  
4   NaN        Q       0-10  
5   E46        S      50-60  
6   NaN        S   

In this example, the 'pd. cut()' function is used to apply binning on the 'Fare' column and create a new column 'Fare_Range'. The 'bins' parameter is used to specify the bin ranges and the 'labels' parameter is used to specify the labels for each bin range. The resulting 'Fare_Range' column will contain the binned values for each passenger's fare.

## Performing statistical measures
Statistical measures are crucial in data aggregation as they simplify the process of summarizing and comprehending large and complex datasets. For instance, the mean, median, and standard deviation of the 'Age' column in the 'passengers_personal_info' dataset can provide insights into the average age of the passengers, and the spread of these values.

In [36]:
# Import the Pandas library
import pandas as pd  

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Calculate the mean of the "Age" column
mean_age = passengers_personal_info["Age"].mean()
print("Mean age:", mean_age)

# Calculate the median of the "Age" column
median_age = passengers_personal_info["Age"].median()
print("Median age:", median_age)

# Calculate the standard deviation of the "Age" column
std_age = passengers_personal_info["Age"].std()
print("Standard deviation of age:", std_age)

Mean age: 24.375
Median age: 24.0
Standard deviation of age: 17.75980614437315


The dataset shows that the passengers on the Titanic had a mean age of 24.375 years, indicating that they were a relatively young population. The median age of 24 years suggests that half of the passengers were 24 years old or younger, while the other half were older than 24. The standard deviation of age was 17.759, indicating that the age distribution was quite spread out.

The 'Fare' column in the 'travel_info' dataset can also be analyzed using mean, median, and standard deviation to understand the average fare and the spread of fares of passengers. This information helps determine the typical fare range and draw conclusions about fare distribution.

In [37]:
# Import the Pandas library
import pandas as pd

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Calculate the mean of the "Fare" column
mean_fare = travel_info["Fare"].mean()
print("Mean fare:", mean_fare)

# Calculate the median of the "Fare" column
median_fare = travel_info["Fare"].median()
print("Median fare:", median_fare)

# Calculate the standard deviation of the "Fare" column
std_fare = travel_info["Fare"].std()
print("Standard deviation of fare:", std_fare)

Mean fare: 22.637488888888885
Median fare: 11.1333
Standard deviation of fare: 23.069552699849016


The dataset shows that the mean fare paid by passengers on the Titanic was 22.64 dollars, indicating that the average fare was relatively low. The median fare of 11.13 dollars suggests that half of the passengers paid less than this amount for their ticket, while the other half paid more. The standard deviation of fare was 23.07 dollars, indicating that there was a wide range of fares paid by passengers, with some paying significantly more or less than the average fare. This variability in fares may reflect differences in ticket class or other factors.

## Performing multi-level indexing
The technique of multilevel indexing is important for aggregating data across multiple dimensions. This means that we can analyze data based on different criteria at the same time. In the Titanic dataset, we use multilevel indexing to group data by both the 'Pclass' and 'Embarked' columns. This allows us to investigate the relationship between these two dimensions and other columns such as 'Age' and 'Fare'. By using multilevel indexing, we can easily compare and contrast different subgroups of the data and see how they relate to each other. 

This can provide valuable insights into patterns and trends in the data that may not be apparent when looking at the data as a whole. Ultimately, multilevel indexing is a powerful tool for exploring complex datasets and uncovering hidden relationships between different variables.

In [38]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Create a multi-level index based on "Pclass" and "Embarked" columns
grouped = titanic.groupby(['Pclass', 'Embarked'])

# Calculate the mean, median, and standard deviation of "Age" and "Fare" columns
result = grouped.agg({'Age': ['mean', 'median', 'std'], 'Fare': ['mean', 'median', 'std']})

# Print the result
print(result)

                       Age                          Fare                    
                      mean median        std        mean   median        std
Pclass Embarked                                                             
1      C         38.027027   36.5  14.243454  104.718529  78.2667  99.093935
       Q         38.500000   38.5   7.778175   90.000000  90.0000   0.000000
       S         38.152037   37.0  15.315584   70.364862  52.0000  58.811278
2      C         22.766667   25.0  10.192551   25.358335  24.0000  11.345067
       Q         43.500000   43.5  19.091883   12.350000  12.3500   0.000000
       S         30.386731   30.0  14.080001   20.327439  13.5000  13.630741
3      C         20.741951   20.0  11.712367   11.214083   7.8958   4.871528
       Q         25.937500   21.5  16.807938   11.183393   7.7500   6.721677
       S         25.696552   25.0  12.110906   14.644083   8.0500  13.276609


In the code above, the 'agg()' method takes a dictionary of columns and aggregation functions, where the keys are the columns to aggregate and the values are the aggregation functions to apply.

The output shows the mean, median, and standard deviation of the 'Age' and 'Fare' columns for each combination of 'Pclass' and 'Embarked'. It can be observed that for 'Pclass' 1, passengers who embarked from Cherbourg (C) paid the highest mean fare of 104.72, whereas those who embarked from Queenstown (Q) paid the highest mean fare of 90.00. Passengers who embarked from Southampton (S) paid the lowest mean fare for the 'Pclass' 1.

# Reshaping Data

## What is reshaping data?
Reshaping data is the process of transforming data from one layout or structure to another. This is a common task in data analysis and involves rearranging data into a more convenient format for analysis, visualization, or storage.

## Why is it important?
Reshaping data is important in data transformation as it allows to change the structure and format of the data into a more suitable or readable form. This can improve the data's overall accessibility, make it easier to perform subsequent analysis or visualization, and help in data-driven decision-making.

One use case for reshaping data is in analyzing customer feedback data. Customer feedback data can come in different formats, such as text, audio, or video, and from various sources, such as social media, email, or surveys. Reshaping this data can help extract valuable insights and trends that may not be apparent from analyzing the raw data.

For example, sentiment analysis is a common technique used to analyze customer feedback data. It involves categorizing customer feedback into positive, negative, or neutral sentiments. Reshaping the data into a structured format, such as a table with sentiment categories and corresponding frequencies, can make it easier to perform sentiment analysis and identify the overall sentiment trends. This information can be used to improve customer satisfaction, identify areas for improvement, and inform marketing strategies.

## Creating pivot tables
A pivot table in Pandas is a powerful tool that allows users to reshape and summarize data in a compact format. With the 'pivot_table()' method, data is aggregated and grouped by one or more variables to create a new, summarized dataframe.

We will pivot the Titanic dataset using the 'Pclass' and 'Embarked' columns as the index, and the 'Fare' column as the values. We chose these columns as the index because they are categorical variables that define different groups of passengers based on their class and port of embarkation, and we want to see how these groups differ in terms of their fare. By using 'Fare' as the value, we can see the average fare for each group of passengers, which allows us to compare the fare across different classes and ports of embarkation.

In [39]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Pivot the data using the 'Pclass' and 'Embarked' columns as the index, and the 'Fare' column as the values
pivot = titanic.pivot_table(index=['Pclass', 'Embarked'], values='Fare')

# Show the result
print(pivot)

                       Fare
Pclass Embarked            
1      C         104.718529
       Q          90.000000
       S          70.364862
2      C          25.358335
       Q          12.350000
       S          20.327439
3      C          11.214083
       Q          11.183393
       S          14.644083


In the code above, the index is based on the 'Pclass' and 'Embarked' columns and the values are based on the 'Fare' column. The resulting pivot table shows the average fare for each combination of 'Pclass' and 'Embarked' values.

Now, we will use the 'Pclass' and 'Sex' columns as the index because they allow us to group passengers based on their class and gender, and we want to analyze how the fare varies between these groups. By using 'Fare' as the values, we can see the average fare for each group of passengers, which helps us compare the fare across different classes and genders.

In [40]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Pivot the data using the 'Pclass' and 'Sex' columns as the index, and the 'Fare' column as the values
pivot_1 = titanic.pivot_table(index=['Pclass', 'Sex'], values='Fare')

# Preview the resulting pivot table
print(pivot_1)

                     Fare
Pclass Sex               
1      female  106.125798
       male     67.226127
2      female   21.970121
       male     19.741782
3      female   16.118810
       male     12.661633


We can also perform multi-level indexing using the 'pivot_table()' method.

In the code below, we selected the 'Pclass' and 'Embarked' columns as the index because they categorize the passengers based on their class and port of embarkation, which can provide insights into how age and fare prices vary for different groups of passengers. We set 'Age' and 'Fare' as the values and applied various aggregation functions to calculate statistics such as mean, median, and standard deviation for both variables.

In [41]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Create a pivot table based on "Pclass" and "Embarked" columns
pivot_table = titanic.pivot_table(index=['Pclass', 'Embarked'], values=['Age', 'Fare'], aggfunc={'Age': ['mean', 'median', 'std'], 'Fare': ['mean', 'median', 'std']})

# Print the pivot table
print(pivot_table)

                       Age                          Fare                    
                      mean median        std        mean   median        std
Pclass Embarked                                                             
1      C         38.027027   36.5  14.243454  104.718529  78.2667  99.093935
       Q         38.500000   38.5   7.778175   90.000000  90.0000   0.000000
       S         38.152037   37.0  15.315584   70.364862  52.0000  58.811278
2      C         22.766667   25.0  10.192551   25.358335  24.0000  11.345067
       Q         43.500000   43.5  19.091883   12.350000  12.3500   0.000000
       S         30.386731   30.0  14.080001   20.327439  13.5000  13.630741
3      C         20.741951   20.0  11.712367   11.214083   7.8958   4.871528
       Q         25.937500   21.5  16.807938   11.183393   7.7500   6.721677
       S         25.696552   25.0  12.110906   14.644083   8.0500  13.276609


The resulting pivot table shows the average age and fare prices, along with their respective statistics, for each combination of 'Pclass' and 'Embarked'. This allows us to compare the age and fare prices across different classes and ports of embarkation.

## Melting data
Melting data is a technique used to restructure a dataset from a wide format to a long format. In the wide format, the dataset has multiple columns for each variable, and each observation is represented by a single row. For example, in a dataset of students' exam scores, each row could represent a single student, and each column could represent the scores for a particular subject. In the long format, each variable is represented by a single column, and the observations are spread across multiple rows. By reshaping the data in this way, it becomes easier to compare and analyze multiple variables.

Comparing survival rates by Gender: You can melt the dataset to have 'PassengerId' as the identifier variable and 'Sex' and 'Survived' as the melted variables. This will give you a long format dataset with a row for each passenger-gender combination, and columns for 'PassengerId', 'Sex', and 'Survived'.

In [42]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Select the column we want to keep as "identifier" variable
id_vars = ["PassengerId"]

# Select the columns we want to melt into the long format
value_vars = ["Sex", "Survived"]

# Melt the dataset to go from a wide format to a long format
df_melted = pd.melt(titanic, id_vars=id_vars, value_vars=value_vars, var_name="variable", value_name="value")

# Print the first few rows of the melted dataset
print(df_melted.head(100))

    PassengerId variable   value
0             1      Sex    male
1             2      Sex  female
2             3      Sex  female
3             4      Sex  female
4             5      Sex    male
..          ...      ...     ...
95           96      Sex    male
96           97      Sex    male
97           98      Sex    male
98           99      Sex  female
99          100      Sex    male

[100 rows x 3 columns]


In the 'pd.melt()' function, the 'id_vars' parameter is used to specify which column(s) to keep as identifier variables, while the 'value_vars' parameter is used to specify which columns to melt into the long format. The 'var_name' and 'value_name' parameters are used to specify the names of the new columns created in the melted dataset. 'var_name' specifies the column containing variable names in the long format, while 'value_name' renames the column containing variable values.

### NOTE

Melting data from a wide format to a long format can be a useful technique for restructuring datasets and making them easier to analyze. However, one potential issue to consider is that melting the data may lead to an increase in the size of the dataset. This is because each row in the original wide format may correspond to multiple rows in the melted long format, resulting in a significant increase in the number of rows. As a result, it is important to be aware of the potential impact on computational efficiency when using this technique.

## Stacking and unstacking data
Stacking and unstacking refer to the process of converting the data representation from one format to another. Stacking is the process of pivoting the data from a wide format to a long format, where each column becomes a single row. Unstacking, on the other hand, is the process of pivoting the data from a long format to a wide format, where each row becomes a single column.

Suppose you want to analyze the relationship between passenger class, port of embarkation, and survival on the Titanic disaster. To do this, you will first group the Titanic data by passenger class, port of embarkation, and survival.


In [43]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Group the data by "Pclass", "Embarked", and "Survived"
grouped_data = titanic.groupby(["Pclass", "Embarked", "Survived"]).size().reset_index(name="count")
print(grouped_data,"\n")

    Pclass Embarked  Survived  count
0        1        C         0     26
1        1        C         1     59
2        1        Q         0      1
3        1        Q         1      1
4        1        S         0     53
5        1        S         1     74
6        2        C         0      8
7        2        C         1      9
8        2        Q         0      1
9        2        Q         1      2
10       2        S         0     88
11       2        S         1     76
12       3        C         0     41
13       3        C         1     25
14       3        Q         0     45
15       3        Q         1     27
16       3        S         0    286
17       3        S         1     67 



In the given code, 'reset_index()' is used to reset the index of the 'grouped_data' DataFrame after it has been grouped and counted by the 'Pclass', 'Embarked', and 'Survived' columns.

Now, the 'pivot_table()' method uses  'Pclass' as the index and 'Embarked' and 'Survived' as the columns. Then the 'stack()' method is used to reshape the pivot table into a long format.

In [44]:
# Import the Pandas library
import pandas as pd

# Load Titanic dataset
titanic = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Group the data by "Pclass", "Embarked", and "Survived"
grouped_data = titanic.groupby(["Pclass", "Embarked", "Survived"]).size().reset_index(name="count")

# Pivot the data using "unstack" to create columns for each embarkation port
pivot_table = grouped_data.pivot_table(index="Pclass", columns=["Embarked", "Survived"], values="count")
print(pivot_table,"\n")

# Use the "stack" method to reshape the data into a long format
# The "reset_index" method is used to reset the index of the reshaped_data DataFrame
reshaped_data = pivot_table.stack().reset_index()

# Print the reshaped data
print(reshaped_data)

Embarked   C       Q        S    
Survived   0   1   0   1    0   1
Pclass                           
1         26  59   1   1   53  74
2          8   9   1   2   88  76
3         41  25  45  27  286  67 

Embarked  Pclass  Survived   C   Q    S
0              1         0  26   1   53
1              1         1  59   1   74
2              2         0   8   1   88
3              2         1   9   2   76
4              3         0  41  45  286
5              3         1  25  27   67


The resulting DataFrame provides a summarized representation of the 'grouped_data' based on the 'Pclass', 'Embarked', and 'Survived' columns.

## Transposing data
In Pandas, transposing a dataframe involves flipping the rows and columns of the original dataframe, so that the rows become columns and the columns become rows. This can be achieved using the '.T' attribute or the 'transpose()' method of the dataframe.

Here is the example using the '.T' attribute:

In [45]:
# Import the Pandas library
import pandas as pd

# Read the "passengers_personal_info.csv" file
passengers_personal_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/passengers_personal_info.csv")

# Transpose the dataframe
titanic_transposed = passengers_personal_info.T

# Preview the transposed dataframe
print(titanic_transposed)

                                   0  \
PassengerId                        1   
Age                             22.0   
Name         Braund, Mr. Owen Harris   
Sex                             male   

                                                             1  \
PassengerId                                                  2   
Age                                                       38.0   
Name         Cumings, Mrs. John Bradley (Florence Briggs Th...   
Sex                                                     female   

                                  2  \
PassengerId                       3   
Age                            26.0   
Name         Heikkinen, Miss. Laina   
Sex                          female   

                                                        3                 4  \
PassengerId                                             4                 6   
Age                                                  35.0               NaN   
Name         Futrelle, Mrs. Jacques

Here is the example using the 'transpose()' method:

In [46]:
# Import the Pandas library
import pandas as pd

# Read the "travel_info.csv" file
travel_info = pd.read_csv("https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/travel_info.csv")

# Transpose the dataframe
titanic_transposed = travel_info.transpose()

# Preview the transposed dataframe
print(titanic_transposed)

                     0         1                 2       3       4        5  \
PassengerId          1         2                 3       5       6        7   
Survived             0         1                 1       0       0        0   
Pclass               3         1                 3       3       3        1   
SibSp                1         1                 0       0       0        0   
Parch                0         0                 0       0       0        0   
Ticket       A/5 21171  PC 17599  STON/O2. 3101282  373450  330877    17463   
Fare              7.25   71.2833             7.925    8.05  8.4583  51.8625   
Cabin              NaN       C85               NaN     NaN     NaN      E46   
Embarked             S         C                 S       S       Q        S   

                  6        7        8  
PassengerId       8        9       11  
Survived          0        1        1  
Pclass            3        3        3  
SibSp             3        0        1  
Parch    

### INFO

The '.T' attribute and 'transpose()' method of a Pandas dataframe can be used interchangeably to transpose the rows and columns of a dataset. However, while '.T' is an attribute of the dataframe object, 'transpose()' is a method that can be called on the dataframe object. Additionally, 'transpose()' allows for additional parameters to be passed, such as whether to preserve the index and column names or reset them.


## Encoding categorical data
Encoding is an essential step in data transformation as it helps to convert categorical variables into a numerical representation that can be processed by machine learning algorithms. Categorical variables are variables that represent discrete values and can take on a limited number of possible values. Examples of categorical variables include gender and education level.

To ensure compatibility with most machine learning algorithms, numerical data is often used as input. Therefore, converting categorical variables into a numerical format is a crucial step. The selection of an appropriate encoding technique depends on the specific type of data and the machine learning algorithm utilized.

Some of the most common encoding techniques include:

* One-hot encoding
* Count encoding
## One-hot encoding
One-hot encoding is a technique that creates a binary column for each category of a categorical variable. This encoding is suitable for nominal categorical variables where the order of categories does not matter. In the Titanic dataset, the 'Sex' and 'Embarked' variables are nominal variables, and we can use one-hot encoding to encode their categories.

Here is the code for performing one-hot encoding on the 'Sex' and 'Embarked' variables in the Titanic dataset using the 'get_dummies()' function:

In [47]:
# Import the Pandas library
import pandas as pd  

# Load the Titanic dataset
titanic_data = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Encode 'Sex' and 'Embarked' variables using One-Hot Encoding
encoded_data = pd.get_dummies(titanic_data, columns=['Sex', 'Embarked'])

# Print the encoded data
print(encoded_data.head(10))

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                             Heikkinen, Miss. Laina  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                           Allen, Mr. William Henry  35.0      0      0   
5                                   Moran, Mr. James   NaN      0      0   
6                            McCarthy, Mr. Timothy J  54.0      0      

In the code above, the 'get_dummies()' function takes the original dataset ('titanic_data') and specifies the columns to be one-hot encoded ('Sex' and 'Embarked') using the 'columns' parameter. It then creates a new dataset ('encoded_data') where each category in the specified columns is replaced with a new binary value.

## Count encoding
Count encoding is a technique that replaces each category of a categorical variable with the count of its occurrences in the dataset. In the Titanic dataset, the 'Embarked' variable is a nominal variable, and we can use count encoding to encode its categories.

Here's the code for count encoding the 'Embarked'  column in the Titanic dataset:

In [48]:
# Import the Pandas library
import pandas as pd  

# Load the Titanic dataset
titanic_data = pd.read_csv('https://staticasssets.blob.core.windows.net/open-ai-coderunner/scripts/titanic.csv')

# Compute the count of each category in the 'Embarked' column
embarked_count = titanic_data.groupby('Embarked').size()

# Map the count values to the corresponding categories
titanic_data['Embarked_count'] = titanic_data['Embarked'].map(embarked_count)

# Print the encoded data
print(titanic_data.head(10))

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54

In this example, we compute the count of each category in the 'Embarked' column using the 'groupby()' method of the Pandas DataFrame. The resulting count values are then mapped to the corresponding categories using the 'map()' method. Finally, we add the resulting count-encoded column 'Embarked_count' to the original dataset.

### NOTE

One-hot encoding creates a new binary value for each category in the variable, resulting in a large number of new features. Count encoding, on the other hand, replaces each category with a single count value. When selecting an encoding technique, it is important to consider the characteristics of the categorical variable and the requirements of the machine learning model.